variables.tf

```diff
@@ -0,0 +1,25 @@
+variable "resource_query" {
```

Comment:
I'm really liking the "single-variable" object pattern. Suggest something like:

```
variable "resource_group" {
  description = ...
  type = object({
    name = string
    type = string
    resource_query = object({
      ResourceTypeFilters = list(string)
      TagFilters = list(object({
        Key    = string,
        Values = list(string)
      }))
    })
  })
```

Reply:
Unfortunately, i don't think we can go with a single-variable approach, findings below;

the `type` is is `required` for a module PRIOR to the resource query
```module "resource_group" {
  source = "../.."
  name   = var.resource_group.name
  type   = var.resource_group.type

  resource_query = {
    ResourceTypeFilters = var.resource_group.resource_query.ResourceTypeFilters
    TagFilters          = var.resource_group.resource_query.TagFilters
  }
  resource_group = var.resource_group
}```

however, for the resource `aws_resourcegroups_group` the `type` is `required` to be part of the `resource_query` so the result if you try to shimmy it; will look like this:
```  # module.resource_group.aws_resourcegroups_group.this will be created
  + resource "aws_resourcegroups_group" "this" {
      + arn      = (known after apply)
      + id       = (known after apply)
      + name     = "ec2_grouped_by_ami"
      + tags_all = (known after apply)

      + resource_query {
          + query = jsonencode(
                {
                  + name           = "ec2_grouped_by_ami"
                  + resource_query = {
                      + ResourceTypeFilters = [
                          + "AWS::EC2::Instance",
                          + "AWS::S3::Bucket",
                        ]
                      + TagFilters          = [
                          + {
                              + Key    = "some_tag_key"
                              + Values = [
                                  + "some_tag_value",
                                ]
                            },
                        ]
                    }
                  + type           = "TAG_FILTERS_1_0"
                }
            )
          + type  = "TAG_FILTERS_1_0"
        }
    }```

Reply:
however, for the resource `aws_resourcegroups_group` the `type` is `required` to be part of the `resource_query` so the result if you try to shimmy it; will look like this:

```  # module.resource_group.aws_resourcegroups_group.this will be created
  + resource "aws_resourcegroups_group" "this" {
      + arn      = (known after apply)
      + id       = (known after apply)
      + name     = "ec2_grouped_by_ami"
      + tags_all = (known after apply)

      + resource_query {
          + query = jsonencode(
                {
                  + name           = "ec2_grouped_by_ami"
                  + resource_query = {
                      + ResourceTypeFilters = [
                          + "AWS::EC2::Instance",
                          + "AWS::S3::Bucket",
                        ]
                      + TagFilters          = [
                          + {
                              + Key    = "some_tag_key"
                              + Values = [
                                  + "some_tag_value",
                                ]
                            },
                        ]
                    }
                  + type           = "TAG_FILTERS_1_0"
                }
            )
          + type  = "TAG_FILTERS_1_0"
        }
    }```

Reply:
also, it looks like when type is part of the ```query = jsonencode() ``` terraform is attempting to infer it's meaning, and places it twice

same is true for `name` input, both need to be a child attribute of `resource`

this may required some more advanced techniques where we have a single-variable, but then we manipulate the JSON for what the api wants, which would then get messy.

Reply:
In the single variable approach, this would be:

```
module "resource_group" {
  source = "../.."
  resource_group = {
    name   = local.resource_group.name
    type   = local.resource_group.type

    resource_query = local.resource_group.resource_query
  }
}
```

In the module, instead of wrapping `jsonencode` around the entire variable, you would specify just the attribute you want to encode, e.g. `jsonencode(var.resource_group.resource_query)`

Reply:
resolved